### PR TITLE
cw20-merkle-airdrop: implement signature check

### PR DESCRIPTION
**Pull request by orkunkl**
Adds new optional signature info field where user can pass signature information to prove identity on another chain.

### CODE REVIEW

File: contracts/cw20-merkle-airdrop/src/state.rs [hunk 1]
```diff
@@ -31,11 +31,19 @@ pub const STAGE_AMOUNT: Map<u8, Uint128> = Map::new(STAGE_AMOUNT_KEY);
...
 pub const MERKLE_ROOT: Map<u8, String> = Map::new(MERKLE_ROOT_PREFIX);
 
 pub const CLAIM_PREFIX: &str = "claim";
-pub const CLAIM: Map<(&Addr, u8), bool> = Map::new(CLAIM_PREFIX);
+pub const CLAIM: Map<(String, u8), bool> = Map::new(CLAIM_PREFIX);
 
 pub const CLAIMED_AMOUNT_PREFIX: &str = "claimed_amount";
 pub const CLAIMED_AMOUNT: Map<(&Addr, u8), bool> = Map::new(CLAIMED_AMOUNT_PREFIX);
+
+pub const HRP_PREFIX: &str = "claimed_amount";
```
Comment (orkunkl) [inline contracts/cw20-merkle-airdrop/src/state.rs:48]:
```suggestion
pub const HRP_PREFIX: &str = "hrp";
```

File: contracts/cw20-merkle-airdrop/src/contract.rs [hunk 2]
```diff
@@ -209,16 +223,36 @@ pub fn execute_claim(
         return Err(ContractError::StageExpired { stage, expiration });
     }
 
+    // if present verify signature and extract external address or use info.sender as proof
+    // If not info.sender is not present in the root, verification will fail
```
Comment (ueco-jb) [inline contracts/cw20-merkle-airdrop/src/contract.rs:227]: Could you please rewrite that comment?

File: contracts/cw20-merkle-airdrop/src/contract.rs [hunk 3]
```diff
@@ -497,6 +536,49 @@ pub fn query_total_claimed(deps: Deps, stage: u8) -> StdResult<TotalClaimedRespo
...
+        host_address,
+        external_address,
+    };
+
+    Ok(resp)
+}
+
+// settings for pagination
+const MAX_LIMIT: u32 = 1000;
+const DEFAULT_LIMIT: u32 = 10;
```
Comment (ueco-jb) [inline contracts/cw20-merkle-airdrop/src/contract.rs:555]: nitpicking: consts are generally put on top of file under imports

File: contracts/cw20-merkle-airdrop/src/helpers.rs [hunk 4]
```diff
@@ -0,0 +1,60 @@
...
+use ripemd::{Digest as RipDigest, Ripemd160};
+use schemars::JsonSchema;
+use serde::{Deserialize, Serialize};
+use sha2::{Digest as ShaDigest, Sha256};
+use std::convert::TryInto;
+
+pub fn verify_cosmos(
+    deps: Deps,
+    claim_msg: &SignedClaimMsg,
+    sig: &CosmosSignature,
```
Comment (ueco-jb) [inline contracts/cw20-merkle-airdrop/src/helpers.rs:14]: You could move that function under `impl CosmosSignature`, like:
```rust
impl CosmosSignature {
    pub fn verify(&self, api: Api, claim_msg: &SignedClaimMsg) -> Result<bool, ContractError> {
```

File: contracts/cw20-merkle-airdrop/src/helpers.rs [hunk 5]
```diff
@@ -0,0 +1,60 @@
...
+}
+
+pub fn verify_external_address(
+    deps: &DepsMut,
+    info: &MessageInfo,
+    hrp: String,
+    sig: &SignatureInfo,
+) -> Result<String, ContractError> {
+    // check if signature is correct
+    let signature: CosmosSignature = from_binary(&sig.signature).unwrap();
```
Comment (ueco-jb) [inline contracts/cw20-merkle-airdrop/src/helpers.rs:41]: You have variables `sig` and `signature` which are completely separate structures.
```suggestion
    let cosmos_signature: CosmosSignature = from_binary(&sig.signature).unwrap();
``` 
?

File: contracts/cw20-merkle-airdrop/src/msg.rs [hunk 6]
```diff
@@ -28,13 +28,19 @@ pub enum ExecuteMsg {
         expiration: Option<Expiration>,
         start: Option<Scheduled>,
         total_amount: Option<Uint128>,
+        // hrp is bech32 prefix. Required for external network airdrop proof verification
```
Comment (ueco-jb) [inline contracts/cw20-merkle-airdrop/src/msg.rs:31]: Could you explain here what `hrp` stands for? I'm missing that bit in whole PR.

File: contracts/cw20-merkle-airdrop/src/msg.rs [hunk 7]
```diff
@@ -85,5 +93,29 @@ pub struct TotalClaimedResponse {
...
+    pub address_maps: Vec<AccountMapResponse>,
+}
+
 #[derive(Serialize, Deserialize, Clone, Debug, PartialEq, JsonSchema)]
 pub struct MigrateMsg {}
+
+// Signature check is done on claim if the value is present in the context
+#[derive(Serialize, Deserialize, Clone, Debug, PartialEq, JsonSchema)]
+pub struct SignatureInfo {
+    /// Owner If None set, contract is frozen.
```
Comment (ueco-jb) [inline contracts/cw20-merkle-airdrop/src/msg.rs:113]: What does that mean? Could you rephrase that?

File: contracts/cw20-merkle-airdrop/src/msg.rs [hunk 8]
```diff
@@ -85,5 +93,29 @@ pub struct TotalClaimedResponse {
...
+
+#[derive(Serialize, Deserialize, Clone, Debug, PartialEq, JsonSchema)]
+pub struct ListAccountMapResponse {
+    pub address_maps: Vec<AccountMapResponse>,
+}
+
 #[derive(Serialize, Deserialize, Clone, Debug, PartialEq, JsonSchema)]
 pub struct MigrateMsg {}
+
+// Signature check is done on claim if the value is present in the context
```
Comment (ueco-jb) [inline contracts/cw20-merkle-airdrop/src/msg.rs:110]: > if the value is present in the context

What/which value?

File: contracts/cw20-merkle-airdrop/src/helpers.rs [hunk 9]
```diff
@@ -0,0 +1,60 @@
...
+    pub pub_key: Binary,
+    pub signature: Binary,
+}
+
+pub fn verify_external_address(
+    deps: &DepsMut,
+    info: &MessageInfo,
+    hrp: String,
+    sig: &SignatureInfo,
+) -> Result<String, ContractError> {
```
Comment (ueco-jb) [inline contracts/cw20-merkle-airdrop/src/helpers.rs:39]: Content of this function goes beyond verification, it also encodes signature as bench32 address.

**Review by ueco-jb — COMMENTED** [review]
Logic seems to checks out.

File: contracts/cw20-merkle-airdrop/src/contract.rs [hunk 10]
```diff
@@ -209,16 +223,46 @@ pub fn execute_claim(
...
+    // if present verify signature and extract external address or use info.sender as proof
+    // if signature is not present in the message, verification will fail since info.sender is not present in the merkle root
+    let proof_addr = match sig_info {
+        None => info.sender.to_string(),
+        Some(sig) => {
+            // verify signature
+            let cosmos_signature: CosmosSignature = from_binary(&sig.signature)?;
+            cosmos_signature.verify(deps.as_ref(), &sig.claim_msg)?;
+
+            // get airdrop stage bech32 prefix and derive proof address from public key.
```
Comment (orkunkl) [inline contracts/cw20-merkle-airdrop/src/contract.rs:235]:
```suggestion
            // get airdrop stage bech32 prefix and derive proof address from public key
```

**Review by ueco-jb — APPROVED** [review]
LGTM
